### PR TITLE
FINERACT-854 Removed string concatenated SQL from CenterReadPlatform

### PR DESCRIPTION
**Pull request by thesmallstar**
Refer: https://issues.apache.org/jira/browse/FINERACT-854 and #725 #723 for background.
The work for this part is completed, but SQLbuilder currently does not support the use of "limit" and "order by" query which I will be adding before this can be merged.

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/group/service/CenterReadPlatformServiceImpl.java
Comment (thesmallstar) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/group/service/CenterReadPlatformServiceImpl.java:136]: Also removing this part is not correct as of NOW, the problem is we are talking SQL query as a query to give users more functionality and certainly that is not supported in SQL builder currently, we probably would need to parse things find relevant queries and then add those to extra criteria still thinking on it. 
Will be looked after limit and order by are supported.

**Comment by thesmallstar** [comment]
Blocked by: https://issues.apache.org/jira/browse/FINERACT-1058

**Comment by vorburger** [comment]
```
CenterIntegrationTest > testListCenters() FAILED
    java.lang.AssertionError: 1 expectation failed.
    Expected status code <200> but was <500>.```

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/group/service/CenterReadPlatformServiceImpl.java
Comment (thesmallstar) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/group/service/CenterReadPlatformServiceImpl.java:136]: @vorburger  WDYT on this?

**Comment by thesmallstar** [comment]
Failed due to https://issues.apache.org/jira/browse/FINERACT-1016
Looking into it.

**Comment by thesmallstar** [comment]
@vorburger @awasum if you like this approach Before we merge this I will add the tests for the same, if not do you have any suggestions?

**Review by vorburger — CHANGES_REQUESTED** [review]
I'm OK with something like this, but have several feedbacks:

1. I don't know if this actually works (PreparedStatement), so will have to trust you that you test that it really does.

2. The current implementation seems to assume that `setLimit()` & Co. would only ever be called after `addCriteria()` - right? That's... not a good idea. Imagine some dumb developer coming along in a few months, not knowing how you implemented this internally (and correctly so, they should not have to). So they do `new SQLBuilder().addOrderBy().addCriteria(...).getSQLTemplate()` - and that would create invalid SQL - agreed?

3. `SQLBuidlerTest` must be extended to cover these new methods.

I'll re-review this week after next (vacation).

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/security/utils/SQLBuilder.java
Comment (vorburger) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/security/utils/SQLBuilder.java]: lower-case _y_:

```suggestion
    public void addOrderBy(Object orderBy) {
```

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/group/service/CenterReadPlatformServiceImpl.java
Comment (vorburger) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/group/service/CenterReadPlatformServiceImpl.java:136]: @thesmallstar I'm not sure I fully understand what you mean here, but starting to think about adding parsing things sounds like the wrong direction - try not to have to do that (it will be a mess - trust me).

**Comment by thesmallstar** [comment]
> 1. I don't know if this actually works (PreparedStatement), so will have to trust you that you test that it really does.

-> Yes it works! I have tested it :) 
> 2. The current implementation seems to assume that `setLimit()` & Co. would only ever be called after `addCriteria()` - right? That's... not a good idea. Imagine some dumb developer coming along in a few months, not knowing how you implemented this internally (and correctly so, they should not have to). So they do `new SQLBuilder().addOrderBy().addCriteria(...).getSQLTemplate()` - and that would create invalid SQL - agreed?

->Agreed I am making the change. 

> 3. `SQLBuidlerTest` must be extended to cover these new methods.

->Extending, I was waiting for the approach to be marked correct :)

**Comment by thesmallstar** [comment]
failed due to connection error. Closing and opening this PR to retest.

**Comment by thesmallstar** [comment]
I am removing the part to add limit and orderby to SQL builder separate from this PR, so as to review this properly, and also we can merge this quickly then and not keep it blocked by the work and testing on that part.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/group/service/CenterReadPlatformServiceImpl.java
Comment (vorburger) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/group/service/CenterReadPlatformServiceImpl.java:136]: @thesmallstar sorry for the huge delay in getting back to you on this one. I've re-read it again now (took me a minute to get back into it). Are you suggesting that we merge this now? Because, unless I misunderstand, this would break the currently existing functionality for these sqlQueryCriteria, agreed? But they really are a problem, huh? I need to dig more into the code to understand where this is coming from..

**Review by vorburger — CHANGES_REQUESTED** [review]

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/group/service/CenterReadPlatformServiceImpl.java
Comment (vorburger) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/group/service/CenterReadPlatformServiceImpl.java:136]: We can't (shouldn't) merge this PR as is, because it will break support for `sqlSearch`. We need to EITHER support it here, OR (my preference) should just cleanly remove it all together - as (now) suggested in https://issues.apache.org/jira/browse/FINERACT-1095.

**Comment by xurror** [comment]
@thesmallstar any updates on this?

**Comment by thesmallstar** [comment]
This build will fail now, will pass only after rebase from #1171

**Comment by thesmallstar** [comment]
@vorburger review/merge this?
Removed SQLsearch
Fixed small error in groups
SQL concatenation remove ;)

**Review by vorburger — APPROVED** [review]
LGTM

**Pull request merged**
